scripts/dirty-check.js

```diff
@@ -0,0 +1,75 @@
+// @ts-check
+
+/**
+ * This script is intended to be run in a GitHub Actions workflow to check if
+ * test execution changed any files under version control. If so, the script
+ * will exit with a non-zero exit code and print an "error" to the GitHub
+ * Actions log for each changed file.
+ *
+ * This is not a shell script because portability.
+ *
+ * @example
+ *
+ * ```yaml
+ * on: [push, pull_request]
+ * jobs:
+ *   test:
+ *     runs-on: ubuntu-latest
+ *   steps:
+ *     - run: your stuff
+ *     - run: node ./scripts/dirty-check.js
+ * ```
+ *
+ * @packageDocumentation
+ * @see
+ * {@link https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions}
+ */
+
+const { promisify } = require('node:util')
+const execFile = promisify(require('node:child_process').execFile)
+
+/**
+ * Asks `git` if any files changed and exits with a non-zero exit code if so.
+ *
+ * @returns {Promise<void>}
+ * @todo We could determine what lines changed in what file so the file would be
+ *   annotated with the error, but it requires some extra parsing of `git diff`
+ *   output.
+ */
+async function main() {
+  await execFile('git', ['add', '.'])
+  const diffResultPromise = execFile(
+    'git',
+    ['diff', '--staged', '-b', '--quiet'],
+    { encoding: 'utf8' },
+  )
+
+  try {
+    await diffResultPromise
+    console.log('::notice::No dirty files detected')
```

Comment:
(could): Nice to have a standard of splitting log messages from parsable output to allow for easy parsing of result. For example for this script: logging to stderr, diff to stdout.
```suggestion
    console.warn('::notice::No dirty files detected')
```
`console.error` and `console.warn` output to stderr whereas `console.log` outputs to stdout.

Reply:
From what I can tell, GitHub only listens on stdout for [commands](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#example-of-a-workflow-command), so this is intentional